src/Symfony/Component/Messenger/HandleTrait.php

```diff
@@ -42,13 +48,18 @@ private function handle(object $message): mixed
         $handledStamps = $envelope->all(HandledStamp::class);
 
         if (!$handledStamps) {
-            throw new LogicException(sprintf('Message of type "%s" was handled zero times. Exactly one handler is expected when using "%s::%s()".', get_debug_type($envelope->getMessage()), static::class, __FUNCTION__));
+            if ($this->allowAsyncHandling && $envelope->all(SentStamp::class)) {
+                // it went through an asynchronous transport
+                return null;
+            }
+
+            throw new NoHandlerForMessageException(sprintf('Message of type "%s" was handled zero times. Exactly one handler is expected when using "%s::%s()".', get_debug_type($envelope->getMessage()), static::class, __FUNCTION__));
```

Comment:
Using a dedicated exception class for this will allow us to catch it and improve the error message for specific message buses; for example, I want to say "Command of type..." instead of "Message of type..." for command buses, and "Query of type..." for query buses.